source/Halibut/Transport/Protocol/MessageExchangeStream.cs

```diff
@@ -246,14 +260,51 @@ void ExpectServerIdentity()
 
         T ReadBsonMessage<T>()
         {
-            using (var zip = new DeflateStream(stream, CompressionMode.Decompress, true))
-            using (var bson = new BsonDataReader(zip) { CloseInput = false })
+            using (var buffer = new BufferedReadStream(stream))
             {
-                var messageEnvelope = serializer.Deserialize<MessageEnvelope>(bson);
-                if (messageEnvelope == null)
-                    throw new Exception("messageEnvelope is null");
-                
-                return (T)messageEnvelope.Message;
+                try
+                {
+                    using (var zip = new DeflateStream(buffer, CompressionMode.Decompress, true))
+                    {
+                        using (var bson = new BsonDataReader(zip) {CloseInput = false})
+                        {
+                            var messageEnvelope = serializer.Deserialize<MessageEnvelope>(bson);
+
+                            log.Write(EventType.Diagnostic, $"ReadBsonMessage: {BitConverter.ToString(buffer.GetBytes())}");
```

Comment:
We want to wrap these log messages in an "Is env var set" type thing. This diagnostics would have been so useful for this. We also want to update the readme for this.
Need to liaise with @OctopusDeploy/school-security to make sure this isn't A Bad Thing :tm: